src/lib/OS/SimOS.cc

```diff
@@ -105,36 +107,44 @@ void SimOS::tick() {
                (scheduledProcs_.front()->status_ == procStatus::completed)) {
           scheduledProcs_.pop();
         }
-        if (!scheduledProcs_.empty()) {
-          // Get context of process that was executing on core before interrupt
-          // was signalled
-          OS::cpuContext currContext = core->getCurrentContext();
-          // Core's stored TID will equal -1 if no process has been previously
-          // scheduled (i.e. on first tick of simulation)
-          if (currContext.TID != -1) {
-            // Find the corresponding process in map
-            auto procItr = processes_.find(currContext.TID);
-            // If proccess can't be found then it has been terminated so no need
-            // to update context.
-            if (procItr != processes_.end()) {
-              auto currProc = procItr->second;
-              assert((currProc->status_ == procStatus::executing) &&
-                     "[SimEng:SimOS] Process updated when not in executing "
-                     "state.");
-              // Only update values which have changed
-              currProc->context_.pc = currContext.pc;
-              currProc->context_.regFile = currContext.regFile;
-              // Change status from Executing to Waiting
-              currProc->status_ = procStatus::waiting;
-              waitingProcs_.push(currProc);
-            }
+        // Get context of process that was executing on core before interrupt
+        // was signalled
+        OS::cpuContext currContext = core->getCurrentContext();
+        // Core's stored TID will equal -1 if no process has been previously
+        // scheduled (i.e. on first tick of simulation)
+        if (currContext.TID != -1) {
+          // Find the corresponding process in map
+          auto procItr = processes_.find(currContext.TID);
+          // If proccess can't be found then it has been terminated so no need
+          // to update context.
+          if (procItr != processes_.end()) {
+            auto currProc = procItr->second;
+            assert((currProc->status_ == procStatus::executing) &&
+                   "[SimEng:SimOS] Process updated when not in executing "
+                   "state.");
+            // Only update values which have changed
+            currProc->context_.pc = currContext.pc;
+            currProc->context_.regFile = currContext.regFile;
+            // Change status from Executing to Waiting
+            currProc->status_ = procStatus::waiting;
+            waitingProcs_.push(currProc);
```

Comment:
Logic looks good, but what if we have only one process, in that case, that process will be pushed on to `waitingProcs_`,  and because of the new optimisation it will be scheduled back again [here](https://github.com/UoB-HPC/SimEng/pull/289/files#diff-23a38459d3c1bce689bdc1777a9d1bcad069b18b4b8870f78193b6ac4d5cc83bR143). Could this somehow be eliminated?

Reply:
If we only have a single process then no context switch will occur due to the test performed in the `CoreStatus::executing` switch case in SimOS.cc